File: NetInfinity/AppState.swift
//
//  AppState.swift
//  NetInfinity
//
//

import SwiftUI
import Combine

// MARK: - App State Management

final class AppState: ObservableObject {
    @Published var authenticationState: AuthenticationState = .unknown
    @Published var currentIdentity: LocalIdentity?
    @Published var colorScheme: ColorScheme = .light
    @Published var isLoading = false
    @Published var error: AppError?
    
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        setupObservers()
    }
    
    private func setupObservers() {
        // Observe authentication state changes
        $authenticationState
            .sink { state in
                print("Authentication state changed to: \(state)")
                // Additional side effects can be added here
            }
            .store(in: &cancellables)
    }
    
    func setReady(with identity: LocalIdentity) {
        currentIdentity = identity
        MessageContext.currentUserId = identity.id
        authenticationState = .ready(identity: identity)
    }
    
    func clearIdentity() {
        currentIdentity = nil
        MessageContext.currentUserId = nil
        authenticationState = .unknown
    }
    
    func setLoading(_ isLoading: Bool) {
        self.isLoading = isLoading
    }
    
    func setError(_ error: AppError?) {
        self.error = error
    }
    
    func clearError() {
        error = nil
    }
}

// MARK: - Authentication State

extension AppState {
    enum AuthenticationState {
        case unknown
        case ready(identity: LocalIdentity)
    }
}

// MARK: - Error Handling

enum AppError: Error, Identifiable {
    case authenticationFailed
    case networkError(Error)
    case invalidSession
    case unknownError
    
    var id: String {
        switch self {
        case .authenticationFailed: return "authenticationFailed"
        case .networkError: return "networkError"
        case .invalidSession: return "invalidSession"
        case .unknownError: return "unknownError"
        }
    }
    
    var title: String {
        switch self {
        case .authenticationFailed: return "Authentication Failed"
        case .networkError: return "Network Error"
        case .invalidSession: return "Invalid Session"
        case .unknownError: return "Error"
        }
    }
    
    var message: String {
        switch self {
        case .authenticationFailed: return "Failed to authenticate. Please try again."
        case .networkError(let error): return "Network error: \(error.localizedDescription)"
        case .invalidSession: return "Your session has expired. Please log in again."
        case .unknownError: return "An unknown error occurred. Please try again."
        }
    }
}

File: frontend/lib/state/app_state.dart
import 'package:flutter/material.dart';

enum AuthenticationState {
  unknown,
  authenticated,
  notAuthenticated,
}

class Session {
  final String userId;
  final String displayName;
  final DateTime createdAt;

  const Session({
    required this.userId,
    required this.displayName,
    required this.createdAt,
  });
}

class AppError {
  final AppErrorType type;
  final String message;
  final dynamic error;

  AppError(this.type, this.message, [this.error]);

  String get title {
    switch (type) {
      case AppErrorType.authenticationFailed:
        return 'Authentication Failed';
      case AppErrorType.networkError:
        return 'Network Error';
      case AppErrorType.invalidSession:
        return 'Invalid Session';
      case AppErrorType.unknownError:
        return 'Error';
    }
  }
}

enum AppErrorType {
  authenticationFailed,
  networkError,
  invalidSession,
  unknownError,
}

class AppState extends ChangeNotifier {
  AuthenticationState _authenticationState = AuthenticationState.unknown;
  Session? _currentSession;
  ThemeMode _themeMode = ThemeMode.light;
  bool _isLoading = false;
  AppError? _error;
  bool _autoSaveMedia = false;
  bool _readReceipts = true;
  bool _autoAcceptTransfers = false;
  bool _notifyOnTransferComplete = true;
  bool _preferLowLatencyRoutes = false;

  AuthenticationState get authenticationState => _authenticationState;
  Session? get currentSession => _currentSession;
  ThemeMode get themeMode => _themeMode;
  bool get isLoading => _isLoading;
  AppError? get error => _error;
  bool get autoSaveMedia => _autoSaveMedia;
  bool get readReceipts => _readReceipts;
  bool get autoAcceptTransfers => _autoAcceptTransfers;
  bool get notifyOnTransferComplete => _notifyOnTransferComplete;
  bool get preferLowLatencyRoutes => _preferLowLatencyRoutes;

  void setAuthenticationState(AuthenticationState state) {
    _authenticationState = state;
    notifyListeners();
  }

  void setCurrentSession(Session? session) {
    _currentSession = session;
    notifyListeners();
  }

  void setThemeMode(ThemeMode mode) {
    _themeMode = mode;
    notifyListeners();
  }

  void setIsLoading(bool loading) {
    _isLoading = loading;
    notifyListeners();
  }

  void setError(AppError? error) {
    _error = error;
    notifyListeners();
  }

  void clearError() {
    _error = null;
    notifyListeners();
  }

  void setAutoSaveMedia(bool value) {
    _autoSaveMedia = value;
    notifyListeners();
  }

  void setReadReceipts(bool value) {
    _readReceipts = value;
    notifyListeners();
  }

  void setAutoAcceptTransfers(bool value) {
    _autoAcceptTransfers = value;
    notifyListeners();
  }

  void setNotifyOnTransferComplete(bool value) {
    _notifyOnTransferComplete = value;
    notifyListeners();
  }

  void setPreferLowLatencyRoutes(bool value) {
    _preferLowLatencyRoutes = value;
    notifyListeners();
  }
}
